// Copyright (C) 2017, Yu Sheng Lin, johnjohnlys@media.ee.ntu.edu.tw

// This file is part of Nicotb.

// Nicotb is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Nicotb is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Nicotb.  If not, see <http://www.gnu.org/licenses/>.
syntax = "proto2";

package NicotbConfig;

message Signal {
	enum NpType { U1 = 1; U8 = 2; U16 = 3; U32 = 4; I8 = 5; I16 = 6; I32 = 7; }
	required string name = 1;
	repeated int32 shape = 2;
	optional NpType np_type = 3 [default = I32];
}

message SignalGroupDefine {
	required string name = 1;
	repeated Signal sigs = 2;
}

message SignalGroup {
	optional string prefix = 1;
	required string grp_def_name = 2;
}

message Bus {
	required string name = 1;
	optional string hier = 2;
	repeated SignalGroup sig_grps = 3;
	repeated Signal sigs = 4;
}

message Event {
	required string name = 1;
	optional string hier = 2;
	repeated string bound_buses = 3;
}

message NicotbConfig {
	repeated SignalGroupDefine siggrp_defs = 1;
	repeated Bus buses = 2;
	repeated Event events = 3;
}
